examples_tests/0.ImportanceSamplingEnvMaps/main.cpp

```diff
@@ -0,0 +1,668 @@
+
+	sceneManager->setActiveCamera(camera);
+
+	auto gpuubo = driver->createDeviceLocalGPUBufferOnDedMem(sizeof(SBasicViewParameters));
+
+	const char* envmapPath = "../../media/envmap/envmap_1.exr";
+	core::smart_refctd_ptr<IGPUImageView> envmapImageView = nullptr;
+	core::smart_refctd_ptr<IGPUImageView> phiPdfLUTImageView = nullptr;
+	core::smart_refctd_ptr<IGPUImageView> thetaLUTImageView = nullptr;
+	{
+		IAssetLoader::SAssetLoadParams lp(0ull, nullptr, IAssetLoader::ECF_DONT_CACHE_REFERENCES);
+		auto envmapImageBundle = assetManager->getAsset(envmapPath, lp);
+		auto envmapImage = core::smart_refctd_ptr_static_cast<asset::ICPUImage>(*envmapImageBundle.getContents().begin());
+		const uint32_t channelCount = getFormatChannelCount(envmapImage->getCreationParameters().format);
+
+		auto luminancePdfBuffer = computeLuminancePdf(envmapImage);
+
+		ICPUImageView::SCreationParams viewParams;
+		viewParams.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0u);
+		viewParams.image = envmapImage;
+		viewParams.format = viewParams.image->getCreationParameters().format;
+		viewParams.viewType = IImageView<ICPUImage>::ET_2D;
+		viewParams.subresourceRange.baseArrayLayer = 0u;
+		viewParams.subresourceRange.layerCount = 1u;
+		viewParams.subresourceRange.baseMipLevel = 0u;
+		viewParams.subresourceRange.levelCount = 1u;
+
+		auto cpuEnvmapImageView = ICPUImageView::create(std::move(viewParams));
+		envmapImageView = driver->getGPUObjectsFromAssets(&cpuEnvmapImageView.get(), &cpuEnvmapImageView.get() + 1u)->front();
+
+		const core::vector2d<uint32_t> pdfDomainExtent = { envmapImage->getCreationParameters().extent.width, envmapImage->getCreationParameters().extent.height };
+
+		core::smart_refctd_ptr<ICPUImage> conditionalCdfImage = nullptr;
+		core::smart_refctd_ptr<ICPUBuffer> conditionalIntegrals = nullptr;
+		{
+			// Create ICPUImage from the buffer for the input image to the SAT filter
+			auto luminanceImageParams = envmapImage->getCreationParameters();
+			luminanceImageParams.format = EF_R32_SFLOAT;
+			luminanceImageParams.extent = { pdfDomainExtent.X, pdfDomainExtent.Y, 1 };
+
+			auto luminanceImageRegions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1ull);
+			luminanceImageRegions->begin()->bufferOffset = 0ull;
+			luminanceImageRegions->begin()->bufferRowLength = luminanceImageParams.extent.width;
+			luminanceImageRegions->begin()->bufferImageHeight = 0u;
+			luminanceImageRegions->begin()->imageSubresource = {};
+			luminanceImageRegions->begin()->imageSubresource.layerCount = 1u;
+			luminanceImageRegions->begin()->imageOffset = { 0, 0, 0 };
+			luminanceImageRegions->begin()->imageExtent = { luminanceImageParams.extent.width, luminanceImageParams.extent.height, 1 };
+
+			core::smart_refctd_ptr<ICPUImage> luminanceImage = ICPUImage::create(std::move(luminanceImageParams));
+			luminanceImage->setBufferAndRegions(core::smart_refctd_ptr(luminancePdfBuffer), luminanceImageRegions);
```

Comment:
if you follow my advice of adding a `CustomPerPixelLambda` to the SAT filter, you'll never need the luma image